Mathlib/RingTheory/QuotSMulTop.lean

```diff
@@ -0,0 +1,131 @@
+/-
+Copyright (c) 2024 Brendan Murphy. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Brendan Murphy
+-/
+import Mathlib.LinearAlgebra.TensorProduct.RightExactness
+
+/-!
+# Reducing a module modulo an element of the ring
+
+Given a commutative ring `R` and an element `r : R`, the association
+`M ↦ M ⧸ rM` extends to a functor on the category of `R`-modules. This functor
+is isomorphic to the functor of tensoring by `R ⧸ (r)` on either side, but can
+be more convenient to work with since we can work with quotient types instead
+of fiddling with simple tensors.
+
+## Tags
+
+module, commutative algebra
+-/
+
+open scoped Pointwise
+
+variable {R} [CommRing R] (r : R) (M : Type*) {M' M''}
+    [AddCommGroup M] [Module R M] [AddCommGroup M'] [Module R M']
+    [AddCommGroup M''] [Module R M'']
+
+/-- An abbreviation for `M⧸rM` that keeps us from having to write
+`(⊤ : Submodule R M)` over and over to satisfy the typechecker. -/
+abbrev QuotSMulTop := M ⧸ r • (⊤ : Submodule R M)
```

Comment:
I considered this. In a previous version of this code I had an smul action of List R on Submodule R M and it made some stuff a lot easier to write down. The downside to what you're suggesting is that R can no longer be inferred from r and M, so we need it as an explicit argument. This isn't a huge issue. The actually awkward thing is that we can't define things like `map` (ie assume the smul action is natural) or assume that smul'ing commutes with suprema. So it's a nice abbreviation and the theory of it works for elements/ideals/sets but (I think) there's no way to unify all those cases with existing typeclasses in mathlib. It didn't feel worth it to add the extra noise of an `R` parameter if all the lemmas in the namespace are about the special case `X = R`. We could add some typeclass support (I'm not entirely sure what this would look like) but I wasn't sure that was a good idea. Thoughts?

Reply:
Hmm, I'm not sure there's a way to set that up so that the two `HasQuotient R (Ideal R)` instances agree. I would love it if we could write `M ⧸ r` though